Add a more fleshed out sample app from volcano code and add zarr reading

SuperOptimizer opened the pull request:
This PR adds in a slightly more fleshed out sample app showing that it

- downloads the zarr .zarray metadata
- parses the metadata
- downloads a zarr block
- decompresses and reads the zarr block
- converts it from uint8 to float32 as a "chunk" which is the internal 3d tensor format I'm using
- generates a mesh using marching cubes
- writes out the mesh as a ply
 
and does this for some volume data from the new standardized zarr and some of @bruniss's fiber predictions.

It has a couple more various fixes and updates:

- adds a cmakelists.txt that detects for cblosc2 and curl and links with them if available
- a few error code fixes (0 is success, != 0 is failure)
- adds the json.c dependency as a git submodule
- adds some backtrace printing code + custom assert to make debugging a little easier
- fixes zarr block reading

The functionality is a bit scattered over a few domains, sorry. If y'all want PRs that are a bit more focused and/or squash (most) everything down to a few commits, I can remake the PR and/or commits, just let me know.

SuperOptimizer committed:
add json.c as a dependency, find cblosc2 in the cmakelists and add a more complete cmakelists.txt

SuperOptimizer committed:
add custom assert with backtrace printing and zarr reading. Fix some tests for new data paths

SuperOptimizer committed:
more involved sample app plus some fixes for the sample app

SuperOptimizer committed:
load both from the volume and the fiber data in sample app

SuperOptimizer committed:
Merge branch 'ScrollPrize:main' into main

ubc-nvining commented on CMakeLists.txt:
```diff
@@ -2,5 +2,47 @@ cmake_minimum_required(VERSION 3.10)
+
+add_compile_options(-g3 -Wall -Wextra)
+if(UNIX)
+    add_link_options(-rdynamic)
```
Is this actually necessary? I'm slightly surprised but it's been awhile.

SuperOptimizer replied:
are you referring to the `-rdynamic` link option? It's not necessary for compiling, but it helps with the `ASSERT` macro for printing the backtrace with function symbols rather than just some virtual address. E.g.


    chunk must be less than 1 gb
    Assertion failed!
    Expression: chunk_dims[0] * chunk_dims[1]*chunk_dims[2]*4 < 1024*1024*1
    Location  : /home/forrest/volcano/third-party/vesuvius-c/vesuvius-c.h:4405
    Function  : vs_vol_get_chunk
    
    Backtrace:
      [0] /home/forrest/volcano/cmake-build-debug/volcano(+0xd219) [0x5628c64f7219]
      [1] /home/forrest/volcano/cmake-build-debug/volcano(+0xd3e6) [0x5628c64f73e6]
      [2] /home/forrest/volcano/cmake-build-debug/volcano(vs_vol_get_chunk+0xb4) [0x5628c6500f4f]
      [3] /home/forrest/volcano/cmake-build-debug/volcano(main+0x80) [0x5628c650583b]
      [4] /lib/x86_64-linux-gnu/libc.so.6(+0x2a1ca) [0x7f76531711ca]
      [5] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x8b) [0x7f765317128b]
      [6] /home/forrest/volcano/cmake-build-debug/volcano(_start+0x25) [0x5628c64ee775]
 
You wouldn't get the `vs_vol_get_chunk` there without `-rdynamic`. Though I need to change this to use `target_link_options` because `add_link_options` _just doesn't work_ and I don't know why

ubc-nvining replied:
OK, add a comment then. I suspect add_link_options vs target_link_options is just one of those cmake things. *shrug*

SuperOptimizer replied:
addressed

ubc-nvining commented on example2.c:
```diff
@@ -2,14 +2,88 @@
+    chunk* mychunk;
+
+    void* metadata_buf;
+    char url[1024] = "https://dl.ash2txt.org/full-scrolls/Scroll1/PHercParis4.volpkg/volumes_zarr_standardized/54keV_7.91um_Scroll1A.zarr/0/.zarray";
```
nit: I suggest [] rather than a predefined size

SuperOptimizer replied:
addressed

ubc-nvining commented on example2.c:
```diff
@@ -2,14 +2,88 @@
+        return 1;
+    }
+    void* compressed_buf;
+    strncpy(url,"https://dl.ash2txt.org/full-scrolls/Scroll1/PHercParis4.volpkg/volumes_zarr_standardized/54keV_7.91um_Scroll1A.zarr/0/50/30/30",1023);
```
... ah, okay, this is why not []. I don't love this; I think I would have two or more strings. Also this way they can be at the start of the example together for if people want to edit one or more at the same time - having it in the middle of the function is a bit weird.

SuperOptimizer replied:
addressed

ubc-nvining commented on example2.c:
```diff
@@ -2,14 +2,88 @@
+    chunk* mychunk;
 
+    void* metadata_buf;
+    char url[1024] = "https://dl.ash2txt.org/community-uploads/bruniss/Fiber-and-Surface-Models/GP-Predictions/updated_zarrs/mask-2ext-surface_ome.zarr/0/.zarray";
```
m.m above

SuperOptimizer replied:
addressed

ubc-nvining commented on vesuvius-c.h:
```diff
@@ -4482,6 +4542,58 @@ zarr_metadata vs_zarr_parse_zarray(char *path) {
   return metadata;
 }
+
+chunk* vs_zarr_read_chunk(char* path, zarr_metadata metadata) {
```
should probably document in the header that the calling function is responsible for cleaning up allocated memory

SuperOptimizer replied:
I'll go confirm, but I think it's always the case that the caller is responsible for `free`ing allocated memory by simply calling `free` if the pointer is a primitive type, otherwise calling `_free` e.g. `chunk_free(chunk* p)` which will free any embedded `malloc`ed code _and_ free the struct pointer itself.

I'll make a note of this in the comment block I have buried in the code and move it to the top of the header.

SuperOptimizer replied:
addressed, see the top of vesuvius-c.h for notes on memory management

ubc-nvining reviewed (commented):
This is basically good and I would approve it as-is, but I think you oughta clean up those string handling things.

SuperOptimizer committed:
- add support for stitching zarr blocks into a single chunk in vs_vol_chunk_get
- add vs_chunk_graft to graft part of a chunk onto another chunk
- add vs_zarr_compress_chunk to be used in the future for writing chunks to disk
- remove vs_fill_chunk and vs_fill_slice as they are superceded by using vs_vol*

SuperOptimizer committed:
remove .gitmodules as we don't use any anymore

ubc-nvining commented on vesuvius-c.h:
```diff
@@ -955,26 +953,30 @@ void reset_mesh_origin_to_roi(TriangleMesh *mesh, const RegionOfInterest *roi) {
+#ifdef MIN
+#undef MIN
 #endif
+#define MIN(a,b) ({__auto_type _a = (a); __auto_type _b = (b); _a < _b ? _a : _b;})
```
This is a GNU specific extension, I believe? I would literally just write it as (a < b ? a : b) and be done with it unless you have a strong argument for otherwise.

SuperOptimizer replied:
yes, statement expressions are a GNU extension. I assume they're supported by clang; dunno about visual studio.

I don't have a strong argument. The only real benefit is that this jumble of code helps avoid multiple evaluations of the parameters if an expression is passed into the macro, and that only really matters if the expression has side effects.

I'll change this to the simple version.

SuperOptimizer replied:
addressed

SuperOptimizer commented on vesuvius-c.h:
```diff
@@ -1758,6 +1812,43 @@ void vs_chunk_set(chunk *chunk, s32 z, s32 y, s32 x, f32 data) {
+        }
+  }
+
+  for (int z = 0; z < dims[2]; z++) {
+    for (int y = 0; y < dims[1]; y++) {
+      for (int x = 0; x < dims[0]; x++) {
+        f32 value = vs_chunk_get(src, src_start[2] + z, src_start[1] + y, src_start[0] + x);
+        vs_chunk_set(dest, dest_start[2] + z, dest_start[1] + y, dest_start[0] + x, value);
+      }
+    }
+  }
+  return 0;
```
the indexing order here is wrong

SuperOptimizer replied:
addressed

SuperOptimizer commented on vesuvius-c.h:
```diff
@@ -4270,77 +4356,99 @@ volume *vs_vol_new(s32 dims[static 3], bool is_zarr, bool is_tif_stack, bool use
-              printf("could not write all data to %s\n",filepath);
+chunk *vs_vol_get_chunk(volume *vol, s32 chunk_pos[static 3], s32 chunk_dims[static 3]) {
+
+  ASSERT(chunk_dims[0] * chunk_dims[1]*chunk_dims[2]*4 < 1024*1024*1, "chunk must be less than 1 gb");
```
delete

SuperOptimizer replied:
addressed

SuperOptimizer commented on vesuvius-c.h:
```diff
@@ -4482,7 +4604,102 @@ zarr_metadata vs_zarr_parse_zarray(char *path) {
+}
+
+int vs_zarr_compress_chunk(chunk* c, zarr_metadata metadata, void** compressed_data) {
+  ASSERT(c->dims[0] == metadata.chunks[0], "zarr block size mismatch with chunk dims");
```
this should return some kind of error rather than ASSERTing

SuperOptimizer replied:
addressed

SuperOptimizer commented on vesuvius-c.h:
```diff
@@ -4546,53 +4763,5 @@ slice *vs_tiff_to_slice(const char *tiffpath, int index) {
 
 
-
-int vs_slice_fill(slice *slice, volume *vol, int start[static 2], int axis) {
```
these methods were left over from some old stuff and are superseded by vs_vol_get_chunk

SuperOptimizer replied:
addressed

SuperOptimizer committed:
reading and writing to/from the zarr chunk cache in volume might be working now

SuperOptimizer committed:
use target_link_options rather than add_link_options because add_link_options doesn't appear to work

SuperOptimizer committed:
close the file pointer and free memory in vs_zarr_write_chunk

SuperOptimizer commented on vesuvius-c.h:
```diff
@@ -152,6 +174,7 @@ static int fetch_metadata(const char *url, char *buffer) {
 
 // Parses the metadata JSON to retrieve chunk sizes and shape
 static int parse_metadata(const char *buffer) {
+#if 0
```
fix before merge 

This branch doesn't have the json fix so I'm doing a bunch of comment shenanigans right now

SuperOptimizer replied:
addressed

SuperOptimizer commented on vesuvius-c.h:
```diff
@@ -1,14 +1,36 @@
 #include <string.h>
 #include <curl/curl.h>
-#include <json-c/json.h>
+//#include <json-c/json.h>
```
fix before merge

SuperOptimizer replied:
addressed

SuperOptimizer commented on vesuvius-c.h:
```diff
@@ -1,14 +1,36 @@
+#define VESUVIUS_IMPL
```
remove before merge

SuperOptimizer replied:
addressed

SuperOptimizer committed:
- block stitching code, aligned to zarr block dimensions, seems to be working
- cleanup example2.c according to review
- cleanup tests according to review and fix a bunch of improper memory handling
- add small logging functonality to vesuvius-c.h
- add more thorough logging to chunk grafting and vs_vol_get_chunk

SuperOptimizer committed:
convert various printfs to the correct LOG statement

SuperOptimizer committed:
fix for uncompressed zarr data

SuperOptimizer committed:
remove json.h include directory and remove blosc + curl status messages since they're mandatory now

SuperOptimizer committed:
remove commented out json code to get ready for merge

SuperOptimizer committed:
typo fix

SuperOptimizer committed:
cmake support for json-c

SuperOptimizer committed:
add a small TODO to vs_vol_get_chunk
